### ISSUE TITLE
uv_mutex_t implementation is recursive on Windows but non-recursive on UNIX?

### ISSUE THREAD

**Issue by dibyendumajumdar** [libuv/libuv#1022]
Hi

I noticed by looking at the code that the uv_mutex_t implementation is recursive on Windows, but on UNIX platforms the attribute PTHREAD_MUTEX_ERRORCHECK is set which means that the mutex will error out if locked recursively. 

I think that supporting recursive locking is preferable in general as it avoids deadlocks when the same thread acquires the mutex multiple times. Is there a reason why the UNIX implementation does not set PTHREAD_MUTEX_RECURSIVE  instead of PTHREAD_MUTEX_ERRORCHECK?

Regards

**Comment by saghul** [comment]
Related (if not the same): https://github.com/saghul/pyuv/issues/216

**Comment by saghul** [comment]
I don't have a strong opinion here. Consistency would be nice though. At any rate I don't think we can make this change in the v1.x series.

**Comment by bnoordhuis** [comment]
Changing uv_mutex_init() so you can opt in to PTHREAD_MUTEX_RECURSIVE might be an acceptable change but making it the default is not.  It's too easy for bugs to go unnoticed that way.

> I think that supporting recursive locking is preferable in general as it avoids deadlocks when the same thread acquires the mutex multiple times.

http://www.zaval.org/resources/library/butenhof1.html is an insightful critique on recursive locks.  I won't try to summarize it here but the main takeaway is that recursive locking is almost always a code smell: if you think you need them, you should stop and rethink what you are doing.

**Comment by dibyendumajumdar** [comment]
Hi

I happen to disagree with the statement that recursive locks are bad. In fact a recursive lock requires the caller to unlock the same number of times as they locked, so it is not true that one loses track of locking scheme etc. It is a very deliberate design choice that make code simpler in many scenarios where the same resource needs to be locked in multiple places.

In any case - whichever way you feel about it - the current implementation is inconsistent. CriticalSections are recursive on Windows, so to make the mutex non-recursive by default you have will to do extra work on to stop it from being recursive.

Regards

**Comment by dibyendumajumdar** [comment]
One way to provide users the option might be to add a new init function - uv_mutex_init_recursive(). This would create a recursive mutex. The existing uv_mutex_lock()/uv_mutex_trylock()/uv_mutex_unlock() functions on Windows would need to be modified to raise an error (abort() ?) if the lock is not recursive and is acquired or released more than once. You will need probably a recursion enabled flag and lock count on Windows.

Something like this maybe:

void uv_mutex_lock(uv_mutex_t* mutex) {
  EnterCriticalSection(mutex);
  if (!mutex->recursive && mutex->lock_count > 0)
    abort(); 
  mutex->lock_count++;
}

void uv_mutex_unlock(uv_mutex_t* mutex) {
  if (!mutex->recursive && mutex->lock_count == 0)
    abort();
  mutex->lock_count--;
  LeaveCriticalSection(mutex);
}

Of course on UNIX a non-recursive mutex deadlocks so it is not exactly same behaviour but probably would be acceptable.

**Comment by bnoordhuis** [comment]
Adding a recursion check on Windows may or may not be a good idea but it's not something we can do in the v1.x branch as it would break ABI (and - arguably - API.)

> Of course on UNIX a non-recursive mutex deadlocks

Not exactly, it depends on the mutex type.  POSIX guarantees deadlock with PTHREAD_MUTEX_NORMAL and an EDEADLK error with PTHREAD_MUTEX_ERRORCHECK but it's undefined behavior with PTHREAD_MUTEX_DEFAULT mutexes.

Some pthreads implementations alias PTHREAD_MUTEX_DEFAULT to PTHREAD_MUTEX_NORMAL but that is not something you can rely on in portable code.

> a recursive lock requires the caller to unlock the same number of times as they locked, so it is not true that one loses track of locking scheme etc. It is a very deliberate design choice that makes code simpler in many scenarios where the same resource needs to be locked in multiple places.

The argument that Butenhof makes (one I happen to agree with) is that you should ask yourself why your program doesn't know if it already owns the lock.  The answer is usually "because it's hard", never "because it's great design."

To be clear, I'm not saying recursive locks don't have their uses but those uses are almost always as band-aids on top of existing code bases where more radical changes would be a herculean effort.

**Comment by dibyendumajumdar** [comment]
Hi

Unfortunately Mr Butenhof does not really justify his claims with real examples.

I think that you need both types of mutexes. I would agree that one should not use a recursive
mutex without a proper design rationale for it. I personally use them in certain scenarios where
code is made cleaner/simpler due to them.

Saying recursive mutexes are bad is like saying reference counting is bad - because after all that's
what they are. When reference goes to zero the lock is released. Like in C++, one would carefully decide
whether to use unique_ptr or shared_ptr - similarly when using mutexes one needs to decide which type
of mutex to use. That is why I think it is a deliberate design choice.

I will give you two real examples where code is made simpler because of recursive locks.

Say you have code that processes transactions between two banks. You need to lock the banks as you 
progress through the transaction. Using a recursive mutex makes the handling of the scenario where
both banks happen to be the same (i.e. it is an internal transaction) the same as when the banks are
different. Coding a special case for when the two banks are the same would be code smell rather than
the opposite.

Another example is say you are using locks in an SQL statement. The user issues one statement
which locks row 1. Then the user takes a savepoint, issues another statement which  locks row 1
and row 2. Now the user rolls back to the savepoint. The lock on row 1 should not be released 
because an instance of it was acquired before the savepoint. But lock on row 2 should be
released. Try designing this without recursive locks. 

In summary I think that the collective wisdom of the industry is that recursive locks are needed.
C++ 11 offers standard mutex and recursive_mutex. POSIX mutexes offer the user a choice. Windows
critical sections support recursion. I think that it would be good if libuv offered the user the choice.

**Comment by saghul** [comment]
I think I'd be ok with us adding recursive mutexes on v2. Nothing is black or white, I've used the myself every now and then.

It could be a `flags` argument to `uv_mutex_init` for instance:

~~~~
// 'simple' mutex
uv_mutex_init(mutex, UV_MUTEX_SIMPLE);
// or the equivalent
uv_mutex_init(mutex, 0);
// recursive
uv_mutex_init(mutex, UV_MUTEX_RECURSIVE);
~~~~

@bnoordhuis  WDTY? Would you be willing to write a patch @dibyendumajumdar ?

**Comment by dibyendumajumdar** [comment]
@saghul hi, happy to submit a patch if there is agreement to proceed. However suggest making the change in a way that existing code does not break. So maybe introduce a new mutex type, or add a new init function if using the current type.

Of course the semantics of existing code would be tightened to say that the mutex is not recursive.

**Comment by saghul** [comment]
@dibyendumajumdar I suggested the change for v2, which will be a major release so we can do breaking changes there. We cannot so this for v1.x because of our stable ABI and backwards compatible API guarantees.

**Comment by hh9527** [comment]
@saghul
when will v2 land (or beta release) ?

**Comment by saghul** [comment]
No ETA as of yet, sorry.
